static/js/popover_menus.js

```diff
@@ -73,7 +73,12 @@ export function any_active() {
 }
 
 function on_show_prep(instance) {
-    $(instance.popper).one("click", instance.hide);
+    $(instance.popper).one("click", (e) => {
+        // Don't do preventDefault since it leads to
+        // direct links in popovers not working.
+        e.stopPropagation();
+        instance.hide();
+    });
     popovers.hide_all_except_sidebars(instance);
 }
```

Comment:
This code seems risky at best and unless I'm misunderstanding it, wrong. If we do need it, we should have an extended comment explaining why this is correct behavior. It seems like the `stopPropagation` is going to break click handlers in popovers that are registered globally (for example, if we wanted to have a `.copy` handler that just copies text on clicking items with a certain CSS class).

Is the goal here to just avoid repeating `e.stopPropagation` in individual click handlers within the popovers? If so, we should just repeat ourselves -- stopping propagation is a thing you should do when the interface you're trying to provide to the rest of the app is "I have fully handled this click, and I don't want anything else to process it". It's a good thing for that to be explicitly right there in the click handler for readability. It also gives us the flexibility to have popovers not close when you click on some items inside them -- see for example the color pickers in the streams popover.

Having that happen by __default__ doesn't make any sense -- this code hasn't processed the click beyond closing the popover (which we do want to explicitly do when we've processed a click on an item in it, but also not something we want to be doing randomly if someone clicks on the popover's border or a disabled element, for example).

Maybe I'm missing something -- the lack of a comment as to what this function's purpose is means I may have guessed wrong about what it's there for.

Reply:
There are elements inside popover that don't have a click handler like `<a href="#stream-settings"` which would require us to do a `stopPropagation()` in the reviewed code above. Not doing so could lead to the new overlays/popovers/compose box being closed before they are opened.

> It seems like the stopPropagation is going to break click handlers in popovers that are registered globally 

That is intentional, since we want the event handlers to be registered to the popover element itself and be destroyed once the popover is hidden.


I can add
```js
        e.stopPropagation();
        instance.hide();
```
inside the click handlers of tippy popovers and the change `on_show_prep` to accommodate for the links with no click handlers?

Reply:
That seems like a high level strategy that we can document clearly -- a generic handler for "navigation links that don't require a click handler and sit the popover". Though if we're doing that, we could have a CSS class on the `a` element to register this behavior, perhaps `navigate_and_close_popover`. (Would that need `stopPropagation` at all?). 

That would protect `a` elements that might have some other purpose (like one with `target=_blank` to open something in a new tab) from having unintended behavior applied to them.

(Generally we want to be making behavior be declared explicitly; being able to grep for `navigate_and_close_popover` to see why something is happening is much easier to read than generic framework code).



> That is intentional, since we want the event handlers to be registered to the popover element itself and be destroyed once the popover is hidden.

That makes sense and is how we should define handlers that we know are scoped to a component. But that feels perhaps different from making any component which is implemented using global handlers break if rendered inside a popover. We haven't done an audit for whether all of our components avoid global handlers, so it seems safer to avoid making design choices that would potentially break any that exist (or that we might add in the future) in a very action-at-a-distance way.

Reply:
Ok, used `navigate_and_close_popover`. See the commit `popover_menus: Keep element click handlers self-contained.`.